src/protocols/multi_party_ecdsa/gg_2020/test.rs

```diff
@@ -150,6 +191,210 @@ fn test_sign_n5_t2_ttag4_corrupt_step7_party24() {
     assert!(&res.err().unwrap().bad_actors[..] == &[1, 3])
 }
 
+//
+// As per page13 https://eprint.iacr.org/2020/540.pdf:
+// This step will:
+// 1. This participant will create a Commitment, Decommitment pair on a scalar
+//    ui and then publish the Commitment part.
+// 2. It will create a Paillier Keypair and publish the public key for that.
+//
+#[cfg(test)]
+fn keygen_stage1(
+    participant: usize,
+) -> (
+    Keys,
+    KeyGenBroadcastMessage1,
+    KeyGenDecommitMessage1,
+    DLogStatement,
+) {
+    // Paillier keys and various other values
+    // party_keys.ek is a secret value and it should be encrypted
+    // using a key that is owned by the participant who creates it. Right now it's plaintext but
+    // this is test.
+    //
+    let party_keys = Keys::create(participant - 1);
+    let (bc1, decom) =
+        party_keys.phase1_broadcast_phase3_proof_of_correct_key_proof_of_correct_h1h2();
+    let h1_h2_N_tilde = bc1.dlog_statement.clone();
+    (party_keys, bc1, decom, h1_h2_N_tilde)
+}
+
+//
+// As per page 13 on https://eprint.iacr.org/2020/540.pdf:
+// 1. Decommit the value obtained in stage1.
```

Comment:
important to note: stage 2 must start only after all messages from stage 1 were delivered (true for all stages )

Reply:
added that comment to all stages.